rust/src/traits/samplers/bernoulli/mod.tex

```diff
@@ -0,0 +1,327 @@
+
+First, we observe that by lines \ref{line:0check} and \ref{line:1check}, if $\texttt{prob} = 0$ then \texttt{sample\_bernoulli} returns \texttt{false}, and if $\texttt{prob} = 1$ then \texttt{sample\_bernoulli} returns \texttt{true} which is correct by definition of a Bernoulli random variable. Otherwise, the variable \texttt{max\_coin\_flips} is computed to be the value $\texttt{T::EXPONENT\_BIAS} + \texttt{T::MANTISSA\_BITS}$ which equals $2^{\ell - 1} - 1 + k$ for any $(k,\ell)$-bit float. Since \texttt{prob} has finite precision, there is some $j$ for which $a_i = 0$ for all $i > j$. For all \emph{normal} $(k,\ell)$-bit floating-point numbers, $j \le 2^{\ell - 1} - 1 + k$ by definition. \texttt{sample\_bernoulli} calls \texttt{sample\_geometric\_buffer} with a buffer of length $\lceil \frac{\texttt{max\_coin\_flips}}{8}\rceil$ bytes (as shown in lines \ref{line:maxcoinflips} and \ref{line:bufferlen}) which returns  $\texttt{None}$ if and only if $I > {8\cdot \lceil \frac{2^{\ell - 1} -1 + k}{8}\rceil}$, where $I \sim Geom(0.5)$. Since $I > j$, this index appears in the \texttt{trailing\_zeroes} part of the binary expansion of \texttt{prob} and should always return \texttt{false}, i.e., $a_I = 0$ for all $I > j$. We can therefore restrict our attention to when \texttt{sample\_geometric\_buffer} returns an index $I \le \texttt{max\_coin\_flips}$ and show that \texttt{sample\_bernoulli} always returns $a_I$.
+
+Next \texttt{sample\_bernoulli} directly computes the number of leading zeroes in the binary expansion of \texttt{prob} to be $\texttt{leading\_zeroes} = \texttt{T::EXPONENT\_BIAS} - 1 - \texttt{raw\_exponent(prob)}$, where \texttt{raw\_exponent(prob)} is the value stored in the $\ell$ bits of the exponent. This value is correct by the specification of a $(k,\ell)$-bit float.  \texttt{sample\_bernoulli} then matches on the value \texttt{first\_heads\_index} corresponding to $I \sim Geom(0.5)$ returned by the function \texttt{sample\_geometric\_buffer}: \\
+
+\noindent\textbf{Case 1} ($\texttt{first\_heads\_index} < \texttt{leading\_zeroes}$). \\
+\noindent For any $I < \texttt{leading\_zeroes}$, it follows that $a_I = 0$ and therefore we should return \texttt{false}. In this case, \texttt{sample\_bernoulli} returns \texttt{false}.\\
+
+\noindent\textbf{Case 2} ($\texttt{first\_heads\_index} == \texttt{leading\_zeroes}$). \\
+\noindent When $\texttt{first\_heads\_index} == \texttt{leading\_zeroes}$, it follows that $a_I$ indexes into the implicit bit by the specification of a $(k,\ell)$-bit float. For all $(k,\ell)$-bit normal floats, the implicit bit equals $1$ and therefore we should return \texttt{true}. In this case, \texttt{sample\_bernoulli} returns \texttt{true}.\\
+
+\noindent\textbf{Case 3} ($ \texttt{leading\_zeroes} < \texttt{first\_heads\_index}  <  \texttt{leading\_zeroes}  + \texttt{T::MANTISSA\_BITS}$). \\
```

Comment:
Your case 3 and 4 are swapped here in the proof. For simplicity of the implementation, I handled the case of indexing into the trailing bits before the case of indexing into the mantissa, because the trailing bits check is simple. Handling mantissa indexing last lets you omit all checks, by process of elimination.